inc/export/class-endpoint.php

```diff
@@ -0,0 +1,323 @@
+	 *
+	 * @return WP_REST_Response|WP_Error Response object.
+	 */
+	public function get_items( WP_REST_Request $request ) {
+		$dates = $this->get_date_query_arguments( $request->get_params()['date'] );
+
+		if ( is_wp_error( $dates ) ) {
+			return $dates;
+		}
+
+		$main_query = [
+			'bool' => [
+				'filter' => [
+					[
+						'term' => [
+							'attributes.blogId.keyword' => get_current_blog_id(),
+						],
+					],
+					[
+						'range' => [
+							'event_timestamp' => [
+								'gte' => $dates['start'],
+								'lt' => $dates['end'],
+							],
+						],
+					],
+				],
+			],
+		];
+
+		// The first query returns just the total number of items
+		// so the results can be sliced later.
+		$total_results = Utils\query( [
+			'query' => $main_query,
+			'size' => 0,
+		] );
+
+		// Now get the paginated results.
+		$per_page = $request->get_param( 'per_page' ) ?? 5000; // Should not be more than 10000.
+		$per_page = max( 1, $per_page );
+        $page = (int) $request->get_params()['page'] ?? 1;
+        $total = $total_results['hits']['total']['value'] ?? $total_results['hits']['total']; // ES 7 compat.
+		$max = (int) ceil( $total / $per_page );
+
+		if ( $max === 0 ) {
+			return rest_ensure_response( [] );
+		}
+
+		// Elasticsearch slices need at least 2 pages.
+		$total_pages = max( $max, 2 );
+
+		// This is the slice ID (similar to a current page).
+		$id = $page - 1;
+
+		if ( $id >= $total_pages ) {
+			return rest_ensure_response( [] );
+		}
+
+		$query = [
+			'query' => $main_query,
+			'slice' => [
+				'id' => $id,
+				'max' => $total_pages,
+			],
+			// Elasticsearch needs this even if slice is set. Let's give it a higher number than per_page.
+			'size' => max( 10000, $per_page + 1000 ),
+		];
+
+		$results = Utils\query( $query, [ 'scroll' => $cache_results_minutes . 'm' ] );
+
+		if ( ! is_array( $results ) ) {
+			return rest_ensure_response( [] );
+        }
+
+        // Extract the event source data only.
+        $hits = wp_list_pluck( $results['hits']['hits'], '_source' );
+
+		$response = rest_ensure_response( $hits );
+
+		$response->header( 'X-WP-Total', (int) $total );
+		$response->header( 'X-WP-TotalPages', $total_pages );
+
+		return $response;
+	}
+
+	/**
+	 * Undocumented function
+	 *
+	 * @param bool $served Whether the request has already been served.
+	 * @param WP_HTTP_ResponseInterface $result Result to send to the client. Usually a WP_REST_Response.
+	 * @param WP_REST_Request $request Request used to generate the response.
+	 * @param WP_REST_Server $server Server instance.
+	 * @return true
+	 */
+	public function pre_serve_request( bool $served, $result, WP_REST_Request $request, WP_REST_Server $server ) : bool {
+		$params = $request->get_params();
+
+		if ( ! preg_match( '#/analytics/v1/export/(?P<date>\d{4}-\d{2}-\d{2})#', $request->get_route() ) || 'GET' !== $request->get_method() ) {
+			return $served;
+		}
+
+		if ( ! isset( $params['format'] ) || 'csv' !== $params['format'] ) {
```

Comment:
does it make sense to read from `Accept` header too?

Reply:
I'm not sure, I thought `Accept` headers could contain multiple mime types. Can do though, shouldn't be hard to add.

Reply:
`Accept` can contain multiple MIME types, and it's a bit of a pain to parse, but it's likely worth doing. [RFC 7231 §5.3.2](https://tools.ietf.org/html/rfc7231#section-5.3.2) is the relevant part.

To summarise from my memory: `Accept` is a comma-separated list of multiple values (as all† HTTP headers are). Each value is specified as `<type>/<subtype>` which can optionally have a wildcard (`*`) as the subtype or type and subtype, and can be followed with a semi-colon and then an arbitrary number of `key=value` parameter pairs separated by semicolons. The main one of those parameters is `q` which indicates the weight (i.e. priority) with 1 being the highest.

i.e. `Accept: text/html, application/xhtml+xml, application/xml;q=0.9, image/webp, */*;q=0.8` is
* Give me `text/html`, `application/xhtml+xml`, or `image/webp` with highest weight (q=1)
* Give me `application/xml` as the fallback, with weight 0.9
* Failing those, give me anything with `q=0.8`

Reply:
OK, added. Seems to work ok. Cheers @rmccue !